modules/serverpod_auth/serverpod_auth_server/lib/src/business/password_hash.dart

```diff
@@ -0,0 +1,294 @@
+      case _Argon2idPasswordHashGenerator.identifier:
+        var [_, _, salt, hash] = passwordHashParts;
+        _hash = hash;
+        _type = _Argon2idPasswordHashGenerator.identifier;
+        _hashGenerator = _Argon2idPasswordHashGenerator(
+          salt: salt,
+          pepper: pepper,
+        );
+        break;
+      case _LegacyToArgon2idPasswordHash.identifier:
+        var [_, _, salt, hash] = passwordHashParts;
+        _hash = hash;
+        _type = _LegacyToArgon2idPasswordHash.identifier;
+        _hashGenerator = _LegacyToArgon2idPasswordHash(
+          legacySalt: legacySalt,
+          salt: salt,
+          email: legacyEmail,
+          pepper: pepper,
+        );
+        break;
+    }
+  }
+
+  /// Returns true if the hash was generated using an outdated algorithm.
+  bool shouldUpdateHash() {
+    return [
+      _LegacyPasswordHashGenerator.identifier,
+      _LegacyToArgon2idPasswordHash.identifier,
+    ].contains(_type);
+  }
+
+  /// Returns true if the hash was generated using the legacy algorithm.
+  bool isLegacyHash() {
+    return _type == _LegacyPasswordHashGenerator.identifier;
+  }
+
+  /// Checks if a password matches the hash.
+  ///
+  /// If the password does not match the hash, the [onValidationFailure] function
+  /// will be called with the hash and the password hash as arguments.
+  bool validate(
+    String password, {
+    void Function(String hash, String passwordHash)? onValidationFailure,
+  }) {
+    var passwordHash = _hashGenerator.generateHash(password);
+
+    if (_hash != passwordHash) {
+      onValidationFailure?.call(passwordHash, _hash);
+      return false;
+    }
+
+    return true;
+  }
+
+  /// Creates a new password hash using the legacy algorithm.
+  ///
+  /// If the [email] parameter is provided, the email will be used as an
+  /// additional salt for the password hash.
+  static String legacyHash(String password, String salt, {String? email}) {
+    return _LegacyPasswordHashGenerator(salt: salt, email: email)
+        .generatePasswordHash(password);
+  }
+
+  /// Creates a Argon2id password hash expecting the passed in password hash to
+  /// be a legacy password hash.
+  ///
+  /// This is used to migrate legacy password hashes to Argon2id password
+  /// hashes.
+  ///
+  /// The [legacySalt] parameter is the salt used in the legacy password hash.
+  ///
+  /// The [salt] parameter should only be used in testing and development.
+  /// If omitted a random salt will be generated which is the recommended way
+  /// to use this function in production.
+  ///
+  /// The [allowUnsecureRandom] parameter can be used to allow unsecure random
+  /// number generation. If set to false, an error will be thrown
+  /// if the platform does not support secure random number generation.
+  static String migratedLegacyToArgon2idHash(
+    String legacyHash, {
+    required String legacySalt,
+    String? salt,
+    String? pepper,
+    bool? allowUnsecureRandom,
```

Comment:
Note that the field defaults to true.

The reason for that is that we are planning to release this as part of a minor/patch version, and since we don't know what environment our users are using I didn't want to introduce hardware constraints without action.

I thought it made sense to default it to true and then change the default to false in Serverpod 2.0 requiring users to configure this manually at that stage.

Reply:
The reason I made it nullable is so that we can provide `null` as input to the method meaning that it is not configured. This gives the benefit that we can read configuration (for example looking up a key in a map). and pass it in directly. Otherwise, we need logic to handle that on the outside.